library/classes/Document.class.php

```diff
@@ -171,12 +216,91 @@ function __construct($id = "")
         $this->encounter_id = 0;
         $this->encounter_check = "";
         $this->encrypted = 0;
+        $this->deleted = 0;
 
         if ($id != "") {
             $this->populate();
         }
     }
 
+    /**
+     * Retrieves all of the categories associated with this document
+     */
+    public function get_categories()
+    {
+        if (empty($this->get_id())) {
+            return [];
+        }
+
+        $categories = "Select `id`, `name`, `value`, `parent`, `lft`, `rght`, `aco_spec` FROM `categories` "
+        . "JOIN `categories_to_documents` `ctd` ON `ctd`.`category_id` = `categories`.`id` "
+        . "WHERE `ctd`.`document_id` = ? ";
+        $resultSet = sqlStatement($categories, [$this->get_id()]);
+        $categories = [];
+        while ($category = sqlGetAssoc($resultSet)) {
+            $categories[] = $category;
+        }
+        return $categories;
+    }
+
+    /**
+     *
+     * @return bool true if the document expiration date has expired
+     */
+    public function has_expired()
+    {
+        if (!empty($this->date_expires)) {
+            $dateTime = DateTime::createFromFormat("Y-m-d H:i:s", $this->date_expires);
+            return $dateTime->getTimestamp() >= time();
+        }
+        return false;
+    }
+
+    /**
+     * Checks whether the passed in $user can access the document or not.  It checks against all of the access
+     * permissions for the categories the document is in.  If there are any categories that the document is tied to
+     * that the owner does NOT have access rights to, the request is denied.  If there are no categories tied to the
+     * document, default access is granted.
+     * @param int|null $user  The user we are checking.  If no user is provided it checks against the currently logged in user
+     * @return bool True if the passed in user or current user can access this document, false otherwise.
+     */
+    public function can_access($user = null)
+    {
+        $categories = $this->get_categories();
+
+        // no categories to prevent access
+        if (empty($categories)) {
+            return true;
+        }
+
+        // verify that we can access every single category this document is tied to
+        foreach ($categories as $category) {
+            if (AclMain::aclCheckAcoSpec($category['aco_spec'], $user) === false) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    /**
+     * Checks if a document has been deleted or not
+     * @return bool true if the document is deleted, false otherwise
+     */
+    public function is_deleted()
+    {
+        return $this->deleted != 0;
+    }
+
+    /**
+     * Handles the deletion of a document
+     */
+    public function process_deleted()
+    {
+        // TODO: @bradymiller is this the correct way in the system to delete documents? Do we remove the underlying file?
```

Comment:
OpenEMR currently deletes the file, but a flag is way better (for logging purposes is very problematic to delete things). So lets definitely go with your mechanism.